File: backend/models/payroll.py
from sqlalchemy import Column, Integer, String, ForeignKey, Date, Float, Text, Enum
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import enum
from app.database import Base


class PayrollStatus(str, enum.Enum):
    DRAFT = "DRAFT"
    PROCESSED = "PROCESSED"
    PAID = "PAID"


class Payroll(Base):
    __tablename__ = "payrolls"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    month = Column(Integer, nullable=False)  # 1-12
    year = Column(Integer, nullable=False)
    pay_period_start = Column(Date, nullable=False)
    pay_period_end = Column(Date, nullable=False)
    
    # Earnings
    basic_salary = Column(Float, default=0, nullable=False)
    house_rent_allowance = Column(Float, default=0, nullable=False)
    leave_travel_allowance = Column(Float, default=0, nullable=False)
    city_allowance = Column(Float, default=0, nullable=False)
    performance_pay = Column(Float, default=0, nullable=False)
    night_shift_allowance = Column(Float, default=0, nullable=False)
    miscellaneous = Column(Float, default=0, nullable=False)
    
    # Deductions
    provident_fund = Column(Float, default=0, nullable=False)
    professional_tax = Column(Float, default=0, nullable=False)
    es_is_deduction = Column(Float, default=0, nullable=False)  # ES/IS deduction
    
    # Calculated fields
    total_earnings = Column(Float, default=0, nullable=False)
    total_deductions = Column(Float, default=0, nullable=False)
    net_salary = Column(Float, default=0, nullable=False)
    
    status = Column(Enum(PayrollStatus), default=PayrollStatus.PROCESSED, nullable=False)
    generated_at = Column(Date, server_default=func.current_date(), nullable=False)
    file_url = Column(String, nullable=True)  # URL to PDF payslip if generated

    # Relationships
    user = relationship("User", back_populates="payrolls")

File: frontend/src/types/payroll.ts
export type PayrollStatus = 'DRAFT' | 'PROCESSED' | 'PAID';

export interface Payroll {
  id: number;
  user_id: number;
  month: number;
  year: number;
  pay_period_start: string;
  pay_period_end: string;
  
  // Earnings
  basic_salary: number;
  house_rent_allowance: number;
  leave_travel_allowance: number;
  city_allowance: number;
  performance_pay: number;
  night_shift_allowance: number;
  miscellaneous: number;
  
  // Deductions
  provident_fund: number;
  professional_tax: number;
  es_is_deduction: number;
  
  // Totals
  total_earnings: number;
  total_deductions: number;
  net_salary: number;
  
  status: PayrollStatus;
  generated_at: string;
  file_url?: string;
}

export interface PayrollSummary {
  payrolls: Payroll[];
  latest_payroll?: Payroll;
  total_earnings_ytd: number;
  total_deductions_ytd: number;
  net_salary_ytd: number;
}
